File: PaintedSkin/src/main/java/org/alee/component/skin/parser/DefaultExecutorBuilder.java
package org.alee.component.skin.parser;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

import org.alee.component.skin.R;
import org.alee.component.skin.executor.BasicViewSkinExecutorFactory;
import org.alee.component.skin.executor.ISkinExecutor;
import org.alee.component.skin.executor.SkinElement;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**********************************************************
 *
 * @author: MY.Liu
 * @date: 2021/2/14
 * @description: xxxx
 *
 *********************************************************/
@RestrictTo(RestrictTo.Scope.LIBRARY)
public final class DefaultExecutorBuilder implements IThemeSkinExecutorBuilder {
    /**
     * 换肤支持的属性 背景
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_BACKGROUND = "background";
    /**
     * 换肤支持的属性 前景色
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_FOREGROUND = "foreground";
    /**
     * 换肤支持的属性 字体颜色
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_TEXT_COLOR = "textColor";
    /**
     * 换肤支持的属性 暗示字体颜色
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_TEXT_COLOR_HINT = "textColorHint";
    /**
     * 换肤支持的属性 选中时高亮背景颜色
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_TEXT_COLOR_HIGH_LIGHT = "textColorHighlight";
    /**
     * 换肤支持的属性 链接的颜色
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_TEXT_COLOR_LINK = "textColorLink";
    /**
     * 换肤支持的属性 进度条背景
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_PROGRESS_DRAWABLE = "progressDrawable";
    /**
     * 换肤支持的属性 ListView分割线
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_LIST_VIEW_DIVIDER = "divider";
    /**
     * 换肤支持的属性 填充内容
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_SRC = "src";
    /**
     * 换肤支持的属性 按钮背景
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public static final String ATTRIBUTE_BUTTON = "button";
    private static final Map<Integer, String> SUPPORT_ATTR = new HashMap<>();

    static {
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_background, ATTRIBUTE_BACKGROUND);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_foreground, ATTRIBUTE_FOREGROUND);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_textColor, ATTRIBUTE_TEXT_COLOR);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_textColorHint, ATTRIBUTE_TEXT_COLOR_HINT);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_textColorHighlight, ATTRIBUTE_TEXT_COLOR_HIGH_LIGHT);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_textColorLink, ATTRIBUTE_TEXT_COLOR_LINK);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_progressDrawable, ATTRIBUTE_PROGRESS_DRAWABLE);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_divider, ATTRIBUTE_LIST_VIEW_DIVIDER);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_src, ATTRIBUTE_SRC);
        SUPPORT_ATTR.put(R.styleable.BasicSupportAttr_android_button, ATTRIBUTE_BUTTON);
    }

    /**
     * 解析支持换肤的属性
     *
     * @param context      {@link Context}
     * @param attributeSet {@link AttributeSet}
     * @return {@link SkinElement}
     */
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    @Override
    public Set<SkinElement> parse(@NonNull Context context, @NonNull AttributeSet attributeSet) {
        TypedArray typedArray = context.obtainStyledAttributes(attributeSet, R.styleable.BasicSupportAttr);
        if (null == typedArray) {
            return null;
        }
        Set<SkinElement> elementSet = new HashSet<>();
        try {
            for (Integer key : SUPPORT_ATTR.keySet()) {
                try {
                    if (typedArray.hasValue(key)) {
                        elementSet.add(new SkinElement(SUPPORT_ATTR.get(key), typedArray.getResourceId(key, -1)));
                    }
                } catch (Throwable ignored) {
                }
            }
        } catch (Throwable ignored) {
        } finally {
            typedArray.recycle();
        }
        return elementSet;
    }

    /**
     * 需要换肤执行器
     *
     * @param view    需要换肤的View
     * @param element 需要执行的元素
     * @return {@link ISkinExecutor}
     */
    @Override
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public ISkinExecutor requireSkinExecutor(@NonNull View view, @NonNull SkinElement element) {
        return BasicViewSkinExecutorFactory.requireSkinExecutor(view, element);
    }

    /**
     * 是否支持属性
     *
     * @param view     View
     * @param attrName 属性名称
     * @return true: 支持
     */
    @Override
    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public boolean isSupportAttr(@NonNull View view, @NonNull String attrName) {
        return SUPPORT_ATTR.containsValue(attrName);
    }
}

File: Library/PaintedSkin-Kotlin/src/main/java/org/alee/component/skin/core/parser/DefaultExecutorBuilder.kt
package org.alee.component.skin.core.parser

import android.content.Context
import android.util.AttributeSet
import android.view.View
import androidx.core.content.res.use
import org.alee.component.skin.R
import org.alee.component.skin.R.styleable
import org.alee.component.skin.constant.DefaultSkinAttribute
import org.alee.component.skin.core.executor.fetchSkinExecutor
import org.alee.component.skin.model.SkinAttribute
import org.alee.component.skin.template.ISkinExecutor
import org.alee.component.skin.template.ISkinExecutorBuilder

/**
 * 框架内默认的执行器构造器
 *
 * <p> 详细描述
 *
 * @author MingYu.Liu
 * created in 2023/9/4
 *
 */
internal class DefaultExecutorBuilder : ISkinExecutorBuilder {

    private companion object {
        private val SUPPORT_ATTR: Map<Int, String> = HashMap<Int, String>().apply {
            // View
            put(R.styleable.BasicSupportAttr_android_background, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_BACKGROUND)
            put(R.styleable.BasicSupportAttr_android_foreground, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_FOREGROUND)
            put(R.styleable.BasicSupportAttr_android_backgroundTint, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_BKG_TINT)
            put(R.styleable.BasicSupportAttr_android_foregroundTint, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_FRG_TINT)
            put(R.styleable.BasicSupportAttr_android_scrollbarThumbHorizontal, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_SCROLLBAR_THUMB_HORIZONTAL)
            put(R.styleable.BasicSupportAttr_android_scrollbarThumbVertical, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_SCROLLBAR_THUMB_VERTICAL)
            put(R.styleable.BasicSupportAttr_android_scrollbarTrackHorizontal, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_SCROLLBAR_TRACK_HORIZONTAL)
            put(R.styleable.BasicSupportAttr_android_scrollbarTrackVertical, DefaultSkinAttribute.ViewAttribute.ATTRIBUTE_SCROLLBAR_TRACK_VERTICAL)
            // TextView
            put(R.styleable.BasicSupportAttr_android_textColor, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_TEXT_COLOR)
            put(R.styleable.BasicSupportAttr_android_textColorHint, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_TEXT_COLOR_HINT)
            put(R.styleable.BasicSupportAttr_android_textColorHighlight, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_TEXT_COLOR_HIGH_LIGHT)
            put(R.styleable.BasicSupportAttr_android_textColorLink, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_TEXT_COLOR_LINK)
            put(R.styleable.BasicSupportAttr_android_shadowColor, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_SHADOW_COLOR)
            put(R.styleable.BasicSupportAttr_android_drawableLeft, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_LEFT)
            put(R.styleable.BasicSupportAttr_android_drawableStart, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_START)
            put(R.styleable.BasicSupportAttr_android_drawableTop, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_TOP)
            put(R.styleable.BasicSupportAttr_android_drawableRight, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_RIGHT)
            put(R.styleable.BasicSupportAttr_android_drawableEnd, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_END)
            put(R.styleable.BasicSupportAttr_android_drawableBottom, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_BOTTOM)
            put(R.styleable.BasicSupportAttr_android_drawableTint, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_TINT)
            put(R.styleable.BasicSupportAttr_drawableLeftCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_LEFT)
            put(R.styleable.BasicSupportAttr_drawableStartCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_START)
            put(R.styleable.BasicSupportAttr_drawableTopCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_TOP)
            put(R.styleable.BasicSupportAttr_drawableRightCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_RIGHT)
            put(R.styleable.BasicSupportAttr_drawableEndCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_END)
            put(R.styleable.BasicSupportAttr_drawableBottomCompat, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_BOTTOM)
            put(R.styleable.BasicSupportAttr_drawableTint, DefaultSkinAttribute.TextViewAttribute.ATTRIBUTE_DRAWABLE_TINT)
            // CompoundButton
            put(R.styleable.BasicSupportAttr_android_button, DefaultSkinAttribute.CompoundButtonAttribute.ATTRIBUTE_BUTTON)
            put(R.styleable.BasicSupportAttr_android_buttonTint, DefaultSkinAttribute.CompoundButtonAttribute.ATTRIBUTE_BUTTON_TINT)
            // ImageView
            put(R.styleable.BasicSupportAttr_android_src, DefaultSkinAttribute.ImageViewAttribute.ATTRIBUTE_SRC)
            put(R.styleable.BasicSupportAttr_android_tint, DefaultSkinAttribute.ImageViewAttribute.ATTRIBUTE_TINT)
            put(R.styleable.BasicSupportAttr_srcCompat, DefaultSkinAttribute.ImageViewAttribute.ATTRIBUTE_SRC)
            put(R.styleable.BasicSupportAttr_tint, DefaultSkinAttribute.ImageViewAttribute.ATTRIBUTE_TINT)
            // ProgressBar
            put(R.styleable.BasicSupportAttr_android_progressDrawable, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_PROGRESS_DRAWABLE)
            put(R.styleable.BasicSupportAttr_android_progressTint, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_PROGRESS_TINT)
            put(R.styleable.BasicSupportAttr_android_progressBackgroundTint, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_PROGRESS_BACKGROUND_TINT)
            put(R.styleable.BasicSupportAttr_android_indeterminateDrawable, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_INDETERMINATE_DRAWABLE)
            put(R.styleable.BasicSupportAttr_android_indeterminateTint, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_INDETERMINATE_TINT)
            put(R.styleable.BasicSupportAttr_android_secondaryProgressTint, DefaultSkinAttribute.ProgressBarAttribute.ATTRIBUTE_SECONDARY_PROGRESS_TINT)
            // AbsListView
            put(R.styleable.BasicSupportAttr_android_cacheColorHint, DefaultSkinAttribute.AbsListViewAttribute.ATTRIBUTE_CACHE_COLOR_HINT)
            put(R.styleable.BasicSupportAttr_android_listSelector, DefaultSkinAttribute.AbsListViewAttribute.ATTRIBUTE_LIST_SELECTOR)
            // ListView
            put(R.styleable.BasicSupportAttr_android_divider, DefaultSkinAttribute.ListViewAttribute.ATTRIBUTE_LIST_VIEW_DIVIDER)
            // ExpandableListView
            put(R.styleable.BasicSupportAttr_android_childDivider, DefaultSkinAttribute.ExpandableListViewAttribute.ATTRIBUTE_CHILD_DIVIDER)
            put(R.styleable.BasicSupportAttr_android_groupIndicator, DefaultSkinAttribute.ExpandableListViewAttribute.ATTRIBUTE_GROUP_INDICATOR)
            put(R.styleable.BasicSupportAttr_android_childIndicator, DefaultSkinAttribute.ExpandableListViewAttribute.ATTRIBUTE_CHILD_INDICATOR)
        }
    }

    /**
     * 实现类需要判断是否可以构建对应属性的换肤执行器
     * @param view View 需要换肤的View
     * @param attrName String 需要换肤的属性名称
     * @return Boolean true: 可以构建 false: 不可以构建
     */
    override fun canBuildExecute(view: View, attrName: String) = SUPPORT_ATTR.containsValue(attrName)

    /**
     * 从AttributeSet中解析出需要换肤的属性
     * @param context Context [Context]
     * @param attributeSet AttributeSet [AttributeSet]
     * @return Set<SkinAttribute>? [SkinAttribute]
     */
    override fun parse(context: Context, attributeSet: AttributeSet): Set<SkinAttribute>? {
        return context.obtainStyledAttributes(attributeSet, styleable.BasicSupportAttr).use { typedArray ->
            SUPPORT_ATTR.filter { typedArray.hasValue(it.key) }
                .map { SkinAttribute(it.value, typedArray.getResourceId(it.key, -1)) }
                .toSet()
        }
    }

    /**
     * 根据需要换肤的属性与View构建换肤执行器
     * @param view View 需要换肤的View
     * @param attribute SkinAttribute 需要换肤的属性
     * @return ISkinExecutor [ISkinExecutor]
     */
    override fun buildSkinExecutor(view: View, attribute: SkinAttribute) = view.fetchSkinExecutor()
}
